versioned/persist/tx/src/main/java/org/projectnessie/versioned/persist/tx/TxConnectionProvider.java

```diff
@@ -0,0 +1,151 @@
+/*
+ * Copyright (C) 2020 Dremio
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.projectnessie.versioned.persist.tx;
+
+import io.agroal.api.AgroalDataSource;
+import io.agroal.api.configuration.AgroalConnectionFactoryConfiguration.TransactionIsolation;
+import io.agroal.api.configuration.supplier.AgroalConnectionFactoryConfigurationSupplier;
+import io.agroal.api.configuration.supplier.AgroalConnectionPoolConfigurationSupplier;
+import io.agroal.api.configuration.supplier.AgroalDataSourceConfigurationSupplier;
+import io.agroal.api.security.NamePrincipal;
+import io.agroal.api.security.SimplePassword;
+import java.sql.Connection;
+import java.sql.ResultSet;
+import java.sql.SQLException;
+import java.sql.Statement;
+import java.text.MessageFormat;
+import java.time.Duration;
+import java.time.temporal.ChronoUnit;
+import java.util.List;
+import java.util.Locale;
+import java.util.Map;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+import javax.sql.DataSource;
+
+public class TxConnectionProvider implements AutoCloseable {
+  public void configure(TxDatabaseAdapterConfig config) {}
+
+  public void setupDatabase(
+      Map<String, List<String>> perTableDDLs,
+      List<String> formatParams,
+      boolean metadataUpperCase,
+      boolean batchDDL,
+      String catalog,
+      String schema) {
+    try (Connection c = this.borrowConnection()) {
+      try (Statement st = c.createStatement()) {
+        Object[] formatParamsArray = formatParams.toArray();
+        Stream<String> ddls =
+            perTableDDLs.entrySet().stream()
+                .filter(e -> !tableExists(c, metadataUpperCase, catalog, schema, e.getKey()))
+                .flatMap(e -> e.getValue().stream())
+                .map(s -> MessageFormat.format(s, formatParamsArray));
+
+        executeDDLs(batchDDL, ddls, st);
+      }
+    } catch (Exception e) {
+      throw new RuntimeException(e);
+    }
+  }
+
+  private void executeDDLs(boolean batchDDL, Stream<String> ddls, Statement st)
+      throws SQLException {
+    if (batchDDL) {
+      String ddl = ddls.map(s -> s + ";").collect(Collectors.joining("\n\n"));
+      if (!ddl.isEmpty()) {
+        ddl = "BEGIN;\n" + ddl + "\nEND TRANSACTION;\n";
+        st.execute(ddl);
+      }
+    } else {
+      List<String> ddlStatements = ddls.collect(Collectors.toList());
+      for (String ddl : ddlStatements) {
```

Comment:
well yes, clearly :-P but why the differentiation? iue why not all batch or all individual?

Reply:
ah :)
Obviously some DBs support it and some don't.
Postgres + Cockroach do support it - in theory ... Especially for Cockroach batch-DDL is much faster (and consistent as well). But yea, it did work w/ earlier Cockroach versions, but sadly not with the most recent versions. It's a relatively new feature there though.

Reply:
so performance and consistency?

Reply:
yea, for DBs that do actually work with it. it's probably not a big thing overall, because nobody will probably recreate the schema that often. but a broken schema isn't nice, so i thought it could be useful.